### PR TITLE
add static_in_const feature gate

### PR DESCRIPTION
**Pull request by llogiq**
also updates tests and deletes the spurious .bk files I inadvertently added last time. 

r? @nikomatsakis

### CODE REVIEW

File: src/librustc_typeck/rscope.rs [hunk 1]
```diff
@@ -213,6 +213,45 @@ impl RegionScope for ElidableRscope {
...
+    fn object_lifetime_default(&self, span: Span) -> Option<ty::Region> {
+        Some(self.base_object_lifetime_default(span))
+    }
+
+    fn base_object_lifetime_default(&self, span: Span) -> ty::Region {
+        if !self.tcx.sess.features.borrow().static_in_const {
+            self.tcx
+                .sess
+                .struct_span_warn(span,
+                                  "This needs a `'static` lifetime or the \
```
Comment (nikomatsakis) [inline src/librustc_typeck/rscope.rs]: Nit: it should be `this` not `This`, and this should be an error :)

File: src/test/compile-fail/rfc1623.rs [hunk 2]
```diff
@@ -7,7 +7,7 @@
 // <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
-
```
Comment (nikomatsakis) [inline src/test/compile-fail/rfc1623.rs]: Nit: can you make another test that specifically uses the feature w/o the feature gate and check that it reports an error

File: src/librustc_typeck/rscope.rs [hunk 3]
```diff
@@ -213,6 +213,45 @@ impl RegionScope for ElidableRscope {
...
+impl<'a, 'gcx: 'a + 'tcx, 'tcx: 'a> RegionScope for StaticRscope<'a, 'gcx, 'tcx> {
+    fn anon_regions(&self,
+                    _span: Span,
+                    count: usize)
+                    -> Result<Vec<ty::Region>, Option<Vec<ElisionFailureInfo>>> {
+        Ok(vec![ty::ReStatic; count])
+    }
+
+    fn object_lifetime_default(&self, span: Span) -> Option<ty::Region> {
+        Some(self.base_object_lifetime_default(span))
```
Comment (nikomatsakis) [inline src/librustc_typeck/rscope.rs]: Ah, I see the problem; these routines should be the same as for elidable rscope:

```rust
    fn object_lifetime_default(&self, span: Span) -> Option<ty::Region> {
        // Per RFC #599, object-lifetimes default to 'static unless
        // overridden by context, and this takes precedence over
        // lifetime elision.
        Some(self.base_object_lifetime_default(span))
    }

    fn base_object_lifetime_default(&self, _span: Span) -> ty::Region {
        ty::ReStatic
    }
```

File: src/test/compile-fail/issue-24446.rs [hunk 4]
```diff
@@ -12,7 +12,7 @@ fn main() {
     static foo: Fn() -> u32 = || -> u32 {
         //~^ ERROR: mismatched types
         //~| ERROR: `std::ops::Fn() -> u32 + 'static: std::marker::Sized` is not satisfied
-
+        //~| WARNING: This needs a `'static` lifetime or the `static_in_const` feature
```
Comment (nikomatsakis) [inline src/test/compile-fail/issue-24446.rs]: These cases seem like bugs -- see my comment above.

File: src/libsyntax/feature_gate.rs [hunk 5]
```diff
@@ -295,6 +295,9 @@ declare_features! (
 
     // Allows untagged unions `union U { ... }`
     (active, untagged_unions, "1.13.0", Some(32836)),
+
```
Comment (durka) [inline src/libsyntax/feature_gate.rs]: > `// If you change this list without updating src/doc/reference.md, @cmr will be sad`

File: src/test/compile-fail/rfc1623.rs (hunk 2 above)
Comment (llogiq) [inline src/test/compile-fail/rfc1623.rs]: Sure thing.

File: src/libsyntax/feature_gate.rs (hunk 5 above)
Comment (llogiq) [inline src/libsyntax/feature_gate.rs]: Thanks for the heads-up, added.

File: src/librustc_typeck/rscope.rs (hunk 3 above)
Comment (llogiq) [inline src/librustc_typeck/rscope.rs]: Should I report the error in `anon_regions(..)` then?

Comment (nikomatsakis) [inline src/librustc_typeck/rscope.rs]: > Should I report the error in anon_regions(..) then?

Yes.

Comment (llogiq) [inline src/librustc_typeck/rscope.rs]: Ok, done.

**Comment by llogiq** [comment]
@chriskrycho I even added a sentence about the feature gate to the reference.

**Comment by chriskrycho** [comment]
@llogiq awesome. I'll see if I can add the docs for the feature itself over the weekend!

**Comment by nikomatsakis** [comment]
@bors r+

Looks good! Thanks @llogiq.

**Comment by llogiq** [comment]
Thank you @nikomatsakis for the awesome mentoring.

**Pull request merged**